common/src/main/java/com/wynntils/core/text/StyledText.java

```diff
@@ -0,0 +1,460 @@
+            Component current = currentPair.key();
+            Style parentStyle = currentPair.value();
+
+            // We use getContents here to get this and only this component's string.
+            String componentString =
+                    MutableComponent.create(current.getContents()).getString();
+
+            List<StyledTextPart> styledTextParts = StyledTextPart.fromComponentWithPossibleFormattingCodes(
+                    componentString, current.getStyle(), null, parentStyle);
+
+            // Only actual styles are inherited, string formatting codes are not
+            Style styleToFollowForChildren = current.getStyle().applyTo(parentStyle);
+
+            List<Pair<Component, Style>> siblingPairs = current.getSiblings().stream()
+                    .map(sibling -> new Pair<>(sibling, styleToFollowForChildren))
+                    .collect(Collectors.toList());
+
+            Collections.reverse(siblingPairs);
+            siblingPairs.forEach(deque::addFirst);
+
+            // Disallow empty parts
+            parts.addAll(
+                    styledTextParts.stream().filter(part -> !part.isEmpty()).toList());
+        }
+
+        return new StyledText(parts, temporaryWorkaround, clickEvents, hoverEvents);
+    }
+
+    public static StyledText fromString(String codedString) {
+        return new StyledText(
+                StyledTextPart.fromComponentWithPossibleFormattingCodes(codedString, Style.EMPTY, null, Style.EMPTY),
+                List.of(),
+                List.of());
+    }
+
+    // We don't want to expose the actual string to the outside world
+    // If you need to do an operation with this string, implement it as a method
+    public String getString(PartStyle.StyleType type) {
+        if (type == PartStyle.StyleType.FULL) {
+            return ComponentUtils.getCoded(temporaryWorkaround);
+        }
+
+        StringBuilder builder = new StringBuilder();
+
+        PartStyle previousStyle = null;
+        for (StyledTextPart part : parts) {
+            builder.append(part.getString(previousStyle, type));
+            previousStyle = part.getPartStyle();
+        }
+
+        return builder.toString();
+    }
+
+    public MutableComponent getComponent() {
+        if (parts.isEmpty()) {
+            return Component.empty();
+        }
+
+        MutableComponent component = parts.get(0).getComponent();
+
+        for (int i = 1; i < parts.size(); i++) {
+            component.append(parts.get(i).getComponent());
+        }
+
+        return component;
+    }
+
+    public static StyledText join(String separator, StyledText... texts) {
+        List<StyledTextPart> parts = new ArrayList<>();
+        List<ClickEvent> clickEvents = new ArrayList<>();
+        List<HoverEvent> hoverEvents = new ArrayList<>();
+
+        final int length = texts.length;
+        for (int i = 0; i < length; i++) {
+            StyledText text = texts[i];
+            parts.addAll(text.parts);
```

Comment:
Note that the StyledTexts are not allowed to be empty here. Maybe you should have a private method for adding parts that checks if the part is empty, and if so, skips adding it. We need to be careful when manipulating the parts list so we uphold this invariant.

Reply:
I will make it so the constructor skips empty parts. That way we can't possibly construct a StyledText with an empty part.